File: templates/problems/16_intervals/05_meeting_rooms_ii/solution.go
package meeting_rooms_ii

import "sort"

// solveMinMeetingRooms uses a sweep line approach. O(n log n) time, O(n) space.
func solveMinMeetingRooms(intervals [][]int) int {
	if len(intervals) == 0 {
		return 0
	}
	starts := make([]int, len(intervals))
	ends := make([]int, len(intervals))
	for i, iv := range intervals {
		starts[i] = iv[0]
		ends[i] = iv[1]
	}
	sort.Ints(starts)
	sort.Ints(ends)

	rooms, maxRooms := 0, 0
	s, e := 0, 0
	for s < len(starts) {
		if starts[s] < ends[e] {
			rooms++
			s++
		} else {
			rooms--
			e++
		}
		if rooms > maxRooms {
			maxRooms = rooms
		}
	}
	return maxRooms
}

File: templates/problems/16_intervals/05_meeting_rooms_ii/typescript/solution.ts
/**
 * Use a sweep line approach. O(n log n) time, O(n) space.
 */
export function solveMinMeetingRooms(intervals: number[][]): number {
    if (intervals.length === 0) {
        return 0;
    }
    const starts = intervals.map(iv => iv[0]).sort((a, b) => a - b);
    const ends = intervals.map(iv => iv[1]).sort((a, b) => a - b);

    let rooms = 0, maxRooms = 0;
    let s = 0, e = 0;
    while (s < starts.length) {
        if (starts[s] < ends[e]) {
            rooms++;
            s++;
        } else {
            rooms--;
            e++;
        }
        if (rooms > maxRooms) {
            maxRooms = rooms;
        }
    }
    return maxRooms;
}
